File: src/main/java/org/stt/persistence/stt/InsertHelper.java
package org.stt.persistence.stt;

import org.stt.model.TimeTrackingItem;
import org.stt.persistence.ItemReader;
import org.stt.persistence.ItemWriter;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;

class InsertHelper {
    private final ItemReader reader;
    private final ItemWriter writer;
    private final TimeTrackingItem itemToInsert;
    private Optional<TimeTrackingItem> lastReadItem;

    InsertHelper(ItemReader reader, ItemWriter writer,
                 TimeTrackingItem itemToInsert) {
        this.itemToInsert = Objects.requireNonNull(itemToInsert);
        this.writer = Objects.requireNonNull(writer);
        this.reader = Objects.requireNonNull(reader);
    }

    void performInsert() {
        copyAllItemsEndingAtOrBeforeItemToInsert();
        lastReadItem.ifPresent(this::adjustEndOfLastItemReadAndWrite);
        writer.write(itemToInsert);
        skipAllItemsCompletelyCoveredByItemToInsert();
        lastReadItem.ifPresent(this::adjustStartOfLastItemReadAndWrite);
        copyRemainingItems();
    }

    private void copyAllItemsEndingAtOrBeforeItemToInsert() {
        copyWhile(item -> item.endsAtOrBefore(itemToInsert.getStart()));
    }

    private void adjustStartOfLastItemReadAndWrite(TimeTrackingItem item) {
        TimeTrackingItem itemToWrite = itemToInsert.getEnd()
                .map(end -> end.isAfter(item.getStart()) ? item.withStart(end) : null)
                .orElse(item);
        writer.write(itemToWrite);
    }

    private void adjustEndOfLastItemReadAndWrite(TimeTrackingItem item) {
        if (item.getStart().isBefore(itemToInsert.getStart())) {
            TimeTrackingItem itemBeforeItemToInsert = item
                    .withEnd(itemToInsert.getStart());
            writer.write(itemBeforeItemToInsert);
        }
    }

    private void copyRemainingItems() {
        copyWhile(item -> true);
    }

    private void skipAllItemsCompletelyCoveredByItemToInsert() {
        Optional<TimeTrackingItem> currentItem = lastReadItem;
        while (currentItem.isPresent() && itemToInsert.endsSameOrAfter(currentItem.get())) {
            currentItem = reader.read();
        }
        lastReadItem = currentItem;
    }

    private void copyWhile(Function<TimeTrackingItem, Boolean> condition) {
        while ((lastReadItem = reader.read()).isPresent()
                && condition.apply(lastReadItem.get())) {
            writer.write(lastReadItem.get());
        }
    }
}

File: src/main/kotlin/org/stt/persistence/stt/InsertHelper.kt
package org.stt.persistence.stt

import org.stt.model.TimeTrackingItem
import org.stt.persistence.ItemReader
import org.stt.persistence.ItemWriter

internal class InsertHelper(val reader: ItemReader,
                            val writer: ItemWriter,
                            val itemToInsert: TimeTrackingItem) {
    private var lastReadItem: TimeTrackingItem? = null

    fun performInsert() {
        copyAllItemsEndingAtOrBeforeItemToInsert()
        lastReadItem?.let { this.adjustEndOfLastItemReadAndWrite(it) }
        writer.write(itemToInsert)
        skipAllItemsCompletelyCoveredByItemToInsert()
        lastReadItem?.let { this.adjustStartOfLastItemReadAndWrite(it) }
        copyRemainingItems()
    }

    private fun copyAllItemsEndingAtOrBeforeItemToInsert() {
        copyWhile { it.endsAtOrBefore(itemToInsert.start) }
    }

    private fun adjustStartOfLastItemReadAndWrite(item: TimeTrackingItem) {
        val itemToWrite = itemToInsert.end?.let { end -> if (end.isAfter(item.start)) item.withStart(end) else null }
                ?: item
        writer.write(itemToWrite)
    }

    private fun adjustEndOfLastItemReadAndWrite(item: TimeTrackingItem) {
        if (item.start.isBefore(itemToInsert.start)) {
            val itemBeforeItemToInsert = item
                    .withEnd(itemToInsert.start)
            writer.write(itemBeforeItemToInsert)
        }
    }

    private fun copyRemainingItems() {
        copyWhile { true }
    }

    private fun skipAllItemsCompletelyCoveredByItemToInsert() {
        var currentItem = lastReadItem
        while (currentItem != null && itemToInsert.endsSameOrAfter(currentItem)) {
            currentItem = reader.read()
        }
        lastReadItem = currentItem
    }

    private fun copyWhile(condition: (TimeTrackingItem) -> Boolean) {
        do {
            lastReadItem = reader.read()
            lastReadItem?.let { if (condition(it)) writer.write(it) else return }
        } while (lastReadItem != null)
    }
}
